### PR TITLE
Add `BlockMetadata` in blockchain with AVR and quorum set

### PR DESCRIPTION
**Pull request by remoun**
* This is a partial cherry-pick of #1823.
* Add `BlockMetadata`, containing AVR and quorum set plus a signature, to `ArchiveBlock` and `BlockData`.
* Add `message blockchain.QuorumSet` with mapping to and from `struct mc_consensus_scp_core::QuorumSet`
* Make `BlockData::new()` require metadata, and update callers.
* Split out `VerificationReport` into a separate `mc-attest-verifier-types` crate.
* Split out `QuorumSet` et al into a separate `mc-consensus-scp-core` crate.
* Add crate `mc_blockchain_test_utils` with helpers for creating metadata.

# Motivation
While this is necessary for block streaming (described in [MCIP 29](https://github.com/mobilecoinfoundation/mcips/blob/main/text/0029-block-streaming.md#extending-blockchain-protobufs)), it's also useful to decouple adding AVRs and quorum sets to the blockchain.

Note that while the `BlockMetadata` can be used to verify chain integrity, it couldn't be used to falsify any new transactions or blocks.

See https://github.com/mobilecoinfoundation/mcips/pull/43 for more details.

# Follow-up work
* #2052
* #2104

### CODE REVIEW

File: transaction/core/src/blockchain/block_metadata.rs
Comment (remoun) [inline transaction/core/src/blockchain/block_metadata.rs:100]: _Originally posted by @iamalwaysuncomfortable in https://github.com/mobilecoinfoundation/mobilecoin/pull/2052#discussion_r881897742_

> I'm unsure whether we should be using `try_sign_digest` or `try_sign` here. 
>
> Specifically:
> - `digest()` makes a hash of a `message` and I'm unsure whether the `digest` should be considered the `message` itself or whether the `digest` should be considered a `pre-hash` of the message? Perhaps we should be signing a specific byte representation of `BlockMetaData` first? Any advice from @isis-mc or @garbageslam would be welcome here.
>- I'm also unsure whether we should have any domain separators added to the digest first. I think we might want that?

File: attest/verifier/types/src/verification.rs
Comment (iamalwaysuncomfortable) [inline attest/verifier/types/src/verification.rs:1]: Same review as other PR. Everyone seems to be moving towards workspace/types pattern - let's discuss if it's something we want to make canonical!

File: ledger/db/src/ledger_db.rs
Comment (iamalwaysuncomfortable) [inline ledger/db/src/ledger_db.rs:261]: Same review: Add tracking issue for FIXMEs

File: transaction/core/src/blockchain/block_metadata.rs
Comment (iamalwaysuncomfortable) [reply transaction/core/src/blockchain/block_metadata.rs:100]: My intuition here is the digest is the correct message (it'd go through another hash, but that's a known set of functions/morphisms to apply) - but maybe a digest with a domain separator. My other intuition is that we might do what's already being done around the codebase as far as "what constitutes the `message` being signed" - but I havne't looked deep enough to know exactly what that is.

File: transaction/core/src/blockchain/quorum_set.rs
Comment (iamalwaysuncomfortable) [inline transaction/core/src/blockchain/quorum_set.rs:57]: @eranrund - any opinions about the duping of QuorumSet here?

Comment (eranrund) [reply transaction/core/src/blockchain/quorum_set.rs:57]: Unavoidable, since the QuorumSet in the scp crate is generic over the node id and can't be used in prost/protobufs :/

**Comment by eranrund** [comment]
Whats the difference between this and https://github.com/mobilecoinfoundation/mobilecoin/pull/2052?

**Comment by remoun** [comment]
> Whats the difference between this and #2052?

I split this out from #2052 to separate the transaction-core-related changes from the ledger-related changes. This PR defines the type and conversions, that PR now builds on this PR, making consensus publish metadata and `LedgerDB` persist it.

File: transaction/core/src/blockchain/block_metadata.rs
Comment (eranrund) [reply transaction/core/src/blockchain/block_metadata.rs:100]: Tagging @isis-mc for a review on this. Thank you

File: transaction/core/src/blockchain/quorum_set.rs
Comment (eranrund) [inline transaction/core/src/blockchain/quorum_set.rs:27]: I am worried that having this definition differ from the one in the `.proto` file is going to cause us headaches in the future. Right now there is a 1:1 mapping between the two and afaik this is the only place in the codebase where the byte representation is not actually compatible.
This will be frustrating to debug if one is handed bytes encoded from the protobuf definition and tries to decode them using the Prost definition.

File: util/from-random/src/lib.rs
Comment (eranrund) [inline util/from-random/src/lib.rs:28]: I don't think these belong here. The purpose of this crate is to just provide the trait.

File: watcher/src/block_data_store.rs
Comment (eranrund) [inline watcher/src/block_data_store.rs:155]: We're gonna want the watcher to store metadata.

**Review by eranrund — CHANGES_REQUESTED** [review]
Overall LGTM but I have some concerns before I can approve:
1. There's the digest signing thing @iamalwaysuncomfortable pointed out
2. The incompatibility between the `QuorumSet` in the `.proto` file and the Prost version of it
3. I believe the watcher should be adapted to store the signed block metadata

Thank you!

**Review by jcape — CHANGES_REQUESTED** [review]
Let's discuss this PR, we should not have duplicate rust definitions of QuorumSet.

File: transaction/core/src/blockchain/block_metadata.rs
Comment (remoun) [reply transaction/core/src/blockchain/block_metadata.rs]: Switched to using `DigestibleSigner`/`DigestibleVerifier`, per this comment and discussion with James. IIUC it's the same operations, with slightly different structure. Definitely could use Isis' review.

File: watcher/src/block_data_store.rs
Comment (remoun) [reply watcher/src/block_data_store.rs:155]: I'm not sure we will. Block streaming will replace watcher and possibly mobilecoind (at least in fog servers), so I don't see the benefit of having watcher track metadata.

Comment (eranrund) [reply watcher/src/block_data_store.rs:155]: I think the standalone watcher is still going to be desirable by people who want to monitor the blockchain for any discrepancies.

File: transaction/core/src/blockchain/quorum_set.rs
Comment (remoun) [reply transaction/core/src/blockchain/quorum_set.rs]: The definition is indeed equivalent. There's a `quorum_set_round_trip` test (in `api/tests/prost.rs`) that verifies this.

I also inspected the code generated by both prost and protobuf, and it is using the same pattern, with the "wrapper" named `QuorumSetMember` and the Rust enum named `QuorumSetMember_oneof_member` (protobuf) or `quorum_set_member::Member` (prost puts it in a `mod`)

File: watcher/src/block_data_store.rs
Comment (remoun) [reply watcher/src/block_data_store.rs:155]: OK, I'll add it

Comment (remoun) [reply watcher/src/block_data_store.rs:155]: Added to #2052

File: util/from-random/src/lib.rs
Comment (eranrund) [reply util/from-random/src/lib.rs:28]: My thinking is that these are only used in tests at the moment, and as such belong in `mc-util-test-helper` but this is a nit not worth spending time on.

Comment (eranrund) [reply util/from-random/src/lib.rs:28]: My thinking is that these are only used in tests at the moment, and as such belong in `mc-util-test-helper` but this is a nit not worth spending time on.

**Review by eranrund — APPROVED** [review]
LGTM pending a resolution of @jcape 's comment `Let's discuss this PR, we should not have duplicate rust definitions of QuorumSet.`. I agree this is desirable, but not sure what the solution is due to the Prost issues.

File: blockchain/types/src/crypto/metadata.rs
Comment (iamalwaysuncomfortable) [reply blockchain/types/src/crypto/metadata.rs:13]: I think I disagree from a Socratic perspective because the signing is a compact amount of code and likely won't deviate from that (or if it did become more complex, I'd think a separate area would then be warranted). However, I'm happy to go along with other opinions if there's a consensus on them.

File: blockchain/types/src/quorum_set.rs
Comment (iamalwaysuncomfortable) [inline blockchain/types/src/quorum_set.rs:28]: I think James held up this PR based on the QuorumSet duplication - should we schedule some time with James to what to do about it?

Comment (remoun) [reply blockchain/types/src/quorum_set.rs:28]: I figured out how to reuse `mc_consensus_scp_core::QuorumSet`

**Comment by eranrund** [comment]
Can you please resolve the conflicts on this one? Thanks!

File: consensus/scp/core/src/quorum_set.rs
Comment (eranrund) [inline consensus/scp/core/src/quorum_set.rs:49]: What is the `1,2` notation? I am not not familiar with that.

**Review by eranrund — APPROVED** [review]

File: api/tests/prost.rs
Comment (remoun) [reply api/tests/prost.rs:19]: This is a test, and only tests opt in to the protobuf feature.

File: consensus/scp/src/slot.rs
Comment (jcape) [reply consensus/scp/src/slot.rs:1948]: Blech, OK

File: blockchain/test-utils/src/lib.rs
Comment (remoun) [reply blockchain/test-utils/src/lib.rs:48]: Filed https://github.com/mobilecoinfoundation/mobilecoin/issues/2184 to track

File: consensus/scp/src/slot.rs
Comment (remoun) [reply consensus/scp/src/slot.rs:1823]: Tracked in https://github.com/mobilecoinfoundation/mobilecoin/issues/1942

File: consensus/service/config/src/network.rs
Comment (remoun) [reply consensus/service/config/src/network.rs:135]: Because of the need for `QuorumSetMemberWrapper`. Suggestions welcome for how to improve this.

**Review by jcape — APPROVED** [review]

**Comment by remoun** [comment]
> I'd like to have a better understanding of the `oneof` thing and the `1,2` tag before this gets merged. Can you please elaborate on how that works and ends up being compatible with the definition in https://github.com/mobilecoinfoundation/mobilecoin/blob/add-block-metadata/api/proto/quorum_set.proto?

The generated code corresponding to `quorum_set.proto` has the same structure:

```rust
#[derive(PartialEq,Clone,Default)]
pub struct QuorumSetMember {
    // message oneof groups
    pub member: ::std::option::Option<QuorumSetMember_oneof_member>,
    // special fields
    pub unknown_fields: ::protobuf::UnknownFields,
    pub cached_size: ::protobuf::CachedSize,
}


#[derive(Clone,PartialEq,Debug)]
pub enum QuorumSetMember_oneof_member {
    node(Node),
    inner_set(QuorumSet),
}

#[derive(PartialEq,Clone,Default)]
pub struct QuorumSet {
    // message fields
    pub threshold: u32,
    pub members: ::protobuf::RepeatedField<QuorumSetMember>,
    // special fields
    pub unknown_fields: ::protobuf::UnknownFields,
    pub cached_size: ::protobuf::CachedSize,
}
```

**Pull request merged**
